File: matrix_class/src/GaussJacobi.cpp
#include "../include/GaussJacobi.hpp"
#include <cmath>
#include <limits>
using namespace std;

GaussJacobi::GaussJacobi() : SystemOfLinearEquationSolver() {}
GaussJacobi::GaussJacobi(int r, int c) : SystemOfLinearEquationSolver(r, c) {}

// Jacobi iterative method
// splits A into D (diagonal) and R (rest)
// x_new[i] = (b[i] - sum(A[i][j]*x_old[j] for j!=i)) / A[i][i]
// uses ONLY old values — stores results in a new array, does NOT overwrite
//
// NO cout here — solver is pure logic. UI layer handles printing.
// Does NOT modify this->data — works on copies.
// Caller is responsible for freeing result.x via delete[].
SolverResult GaussJacobi::solve(double *b, int n, int maxIter, double tol) {
  if (rows != cols || rows != n)
    throw MatrixException("matrix dimensions are sus for Gauss-Jacobi");

  SolverResult result;
  result.n = n;
  result.iterations = 0;
  result.converged = false;
  result.error = 0.0;
  result.dominanceAchieved = true;

  // COPY matrix data and b so we don't mutate the original
  double **A = new double *[n];
  for (int i = 0; i < n; i++) {
    A[i] = new double[n];
    for (int j = 0; j < n; j++)
      A[i][j] = data[i][j];
  }
  double *rhs = new double[n];
  for (int i = 0; i < n; i++)
    rhs[i] = b[i];

  // check diagonal dominance — if not, try to fix by row swapping on copies
  // uses the consolidated base class helper (issue #1)
  // returns bool so we know if dominance was achieved (issue #2, #3)
  if (!isDiagonallyDominant()) {
    result.dominanceAchieved = makeDiagDominant(A, rhs, n);
  }

  // check for zero diagonals
  for (int i = 0; i < n; i++) {
    if (fabs(A[i][i]) < 1e-12) {
      // cleanup everything before throwing (issue #6)
      for (int k = 0; k < n; k++)
        delete[] A[k];
      delete[] A;
      delete[] rhs;
      throw MatrixException(
          "zero on diagonal... Jacobi can't work with this");
    }
  }

  // initial guess x = 0
  double *x_old = new double[n];
  double *x_new = new double[n];
  for (int i = 0; i < n; i++) {
    x_old[i] = 0.0;
    x_new[i] = 0.0;
  }

  // iterate
  double lastMaxDiff = 0.0; // track last diff for result.error (issue #4)
  int iter;
  for (iter = 0; iter < maxIter; iter++) {
    bool diverged = false;

    // compute ALL x_new from x_old (use only old values!)
    for (int i = 0; i < n; i++) {
      double sum = 0.0;
      for (int j = 0; j < n; j++) {
        if (j != i)
          sum += A[i][j] * x_old[j];
      }
      x_new[i] = (rhs[i] - sum) / A[i][i];

      // check for NaN/Inf divergence
      if (isnan(x_new[i]) || isinf(x_new[i])) {
        diverged = true;
        break;
      }
    }

    if (diverged) {
      // set error to infinity on divergence (issue #5)
      result.error = numeric_limits<double>::infinity();
      result.x = x_new;
      result.iterations = iter + 1;
      result.converged = false;
      // cleanup all heap allocations (issue #6)
      delete[] x_old;
      for (int i = 0; i < n; i++)
        delete[] A[i];
      delete[] A;
      delete[] rhs;
      return result;
    }

    // check convergence
    double maxDiff = 0.0;
    for (int i = 0; i < n; i++) {
      double diff = fabs(x_new[i] - x_old[i]);
      if (diff > maxDiff)
        maxDiff = diff;
    }
    lastMaxDiff = maxDiff; // always track (issue #4)

    // copy new to old (replace old vector)
    for (int i = 0; i < n; i++)
      x_old[i] = x_new[i];

    if (maxDiff < tol) {
      result.iterations = iter + 1;
      result.converged = true;
      result.error = maxDiff;
      break;
    }
  }

  if (!result.converged) {
    result.iterations = maxIter;
    result.error = lastMaxDiff; // non-zero error on non-convergence (issue #4)
  }

  result.x = x_new;

  // cleanup
  delete[] x_old;
  for (int i = 0; i < n; i++)
    delete[] A[i];
  delete[] A;
  delete[] rhs;

  return result;
}

File: Extended_in_python/pynumerics/solvers/gauss_jacobi.py
"""
Gauss-Jacobi iterative solver.

Ported from C++ GaussJacobi class. Solves Ax = b iteratively using
only old values at each step. Convergence requires diagonal dominance.
"""

import math
from pynumerics.solvers.base import SystemOfLinearEquationSolver
from pynumerics.solvers.result import SolverResult
from pynumerics.exceptions import MatrixException


class GaussJacobi(SystemOfLinearEquationSolver):
    """Gauss-Jacobi iterative solver for Ax = b.

    Splits A into D (diagonal) and R (remainder).
    x_new[i] = (b[i] - sum(A[i][j]*x_old[j] for j!=i)) / A[i][i]
    Uses ONLY old values per iteration — does NOT overwrite in-place.
    """

    def solve(self, b: list[float], max_iter: int = 10000, tol: float = 1e-10) -> SolverResult:
        """Solve Ax = b using Jacobi iteration.

        Args:
            b: Right-hand side vector.
            max_iter: Maximum number of iterations allowed.
            tol: Convergence tolerance (max absolute difference between iterations).

        Returns:
            SolverResult with solution vector and convergence metrics.
        """
        n = len(b)
        if self.rows != self.cols or self.rows != n:
            raise MatrixException("matrix dimensions are sus for Gauss-Jacobi")

        result = SolverResult(n=n, iterations=0, converged=False, error=0.0, dominance_achieved=True)

        # Work on copies so the original matrix/b are not mutated
        A = [row[:] for row in self._data]
        rhs = list(b)

        # Try to enforce diagonal dominance via row swaps
        if not self.is_diagonally_dominant():
            result.dominance_achieved = SystemOfLinearEquationSolver.make_diag_dominant(A, rhs)

        # Check for zero diagonals after potential reordering
        for i in range(n):
            if abs(A[i][i]) < 1e-12:
                raise MatrixException("zero on diagonal... Jacobi can't work with this")

        # Initial guess: x = 0
        x_old = [0.0] * n
        x_new = [0.0] * n

        last_max_diff = 0.0

        for iteration in range(max_iter):
            diverged = False

            # Compute ALL x_new from x_old (use only old values)
            for i in range(n):
                s = sum(A[i][j] * x_old[j] for j in range(n) if j != i)
                x_new[i] = (rhs[i] - s) / A[i][i]

                # Check for NaN/Inf divergence
                if math.isnan(x_new[i]) or math.isinf(x_new[i]):
                    diverged = True
                    break

            if diverged:
                result.error = float('inf')
                result.x = list(x_new)
                result.iterations = iteration + 1
                result.converged = False
                return result

            # Check convergence
            max_diff = max(abs(x_new[i] - x_old[i]) for i in range(n))
            last_max_diff = max_diff

            # Copy new → old
            x_old = list(x_new)

            if max_diff < tol:
                result.iterations = iteration + 1
                result.converged = True
                result.error = max_diff
                break

        if not result.converged:
            result.iterations = max_iter
            result.error = last_max_diff

        result.x = list(x_new)
        return result
